### PR TITLE
MINOR; Fix nanosecond elapsed time

### PR DESCRIPTION
**Pull request by jsancio**
`Time.nanoseconds()` (`System.nanoTime()`) is not monotonically increasing and it is allowed for the value to overflow (wrap). It can only be used to measure elapsed time. Elapsed time can be computed by subtracting two nanoseconds() calls.

I ran `testBalancePartitionLeaders` 1000 times using jqwik without any failures.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java [hunk 1]
```diff
@@ -445,28 +444,24 @@ public void testBalancePartitionLeaders() throws Throwable {
...
 
-            AtomicLong lastHeartbeatMs = new AtomicLong(getMonotonicMs(active.time()));
+            AtomicLong lastHeartbeatNs = new AtomicLong(active.time().nanoseconds());
             sendBrokerHeartbeat(active, allBrokers, brokerEpochs);
             // Check that partitions are balanced
             TestUtils.waitForCondition(
                 () -> {
-                    long currentMonotonicMs = getMonotonicMs(active.time());
-                    if (currentMonotonicMs > lastHeartbeatMs.get() + (sessionTimeoutMillis / 2)) {
-                        lastHeartbeatMs.set(currentMonotonicMs);
```
Comment (jsancio) [inline metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:455]: If `active.time().nanoseconds()` overflows (wraps), `currentMonotonicMs > lastHeartbeatMs.get() + (sessionTimeoutMillis / 2)` will be false until roughly 292 years (2^63 nanoseconds). The Java Doc recommends to not comparing `nanoTime()` and to instead compare elapsed time `nanoTime() - nanoTime()`. See the Java Doc for details: https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#nanoTime--

**Comment by divijvaidya** [comment]
Wow! This is news to me. BTW there is a JIRA tracking this test failure at https://issues.apache.org/jira/browse/KAFKA-15052 and you might want to associate this PR with that JIRA.

We uses System.nanoTime() for comparison all over the place such as at https://github.com/apache/kafka/blob/f23394336a7741bf4eb23fcde951af0a23a69bd0/core/src/main/scala/kafka/server/metadata/ZkMetadataCache.scala#L573, https://github.com/apache/kafka/blob/f23394336a7741bf4eb23fcde951af0a23a69bd0/server-common/src/main/java/org/apache/kafka/queue/KafkaEventQueue.java#L231 and many many other places. Do you think those places may be having the same overflow problem? If yes, can I request you to please create a JIRA for it?

**Comment by divijvaidya** [comment]
`testBalancePartitionLeaders` is still failing in the CI after this change. Can you please check that?

**Comment by jsancio** [comment]
> `testBalancePartitionLeaders` is still failing in the CI after this change. Can you please check that?

Correct. I'll try to spend sometime this week to figure out why it is failing.

**Pull request closed without merging**